### PR TITLE
Fix post editor layout when content block has no attributes.

### PR DESCRIPTION
**Pull request by tellthemachines**
Trac ticket: https://core.trac.wordpress.org/ticket/59358

This PR syncs the changes from https://github.com/WordPress/gutenberg/pull/54485 and adds a test for the Post Content block with empty attributes.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/block-editor.php [hunk 1]
```diff
@@ -635,7 +635,7 @@ function get_block_editor_settings( array $custom_settings, $block_editor_contex
 
 	$post_content_block_attributes = wp_get_post_content_block_attributes();
 
-	if ( ! empty( $post_content_block_attributes ) ) {
+	if ( isset( $post_content_block_attributes ) ) {
```
Comment (felixarntz) [inline src/wp-includes/block-editor.php:639]: I'm not sure this change is the right one. Now that it's using `isset`, it would also enter the condition if no attributes are present on the post content block, since the default return value is an array.

File: tests/phpunit/tests/blocks/editor.php [hunk 2]
```diff
@@ -454,6 +454,14 @@ public function test_wp_get_post_content_block_attributes() {
 		$this->assertSame( $attributes_with_layout, wp_get_post_content_block_attributes() );
 	}
 
+	public function test_wp_get_post_content_block_attributes_no_layout() {
+		$attributes_empty = array();
+
+		switch_theme( 'block-theme-post-content-default' );
+
+		$this->assertSame( $attributes_empty, wp_get_post_content_block_attributes() );
+	}
```
Comment (felixarntz) [inline tests/phpunit/tests/blocks/editor.php:461]: Related to my above feedback, maybe add another test for `get_block_editor_settings()`, to check that it doesn't include the `postContentAttributes` field if the post block attributes aren't set.

File: tests/phpunit/tests/blocks/editor.php [hunk 3]
```diff
@@ -454,6 +454,14 @@ public function test_wp_get_post_content_block_attributes() {
 		$this->assertSame( $attributes_with_layout, wp_get_post_content_block_attributes() );
 	}
 
+	public function test_wp_get_post_content_block_attributes_no_layout() {
+		$attributes_empty = array();
+
+		switch_theme( 'block-theme-post-content-default' );
+
+		$this->assertSame( $attributes_empty, wp_get_post_content_block_attributes() );
```
Comment (felixarntz) [inline tests/phpunit/tests/blocks/editor.php]: Nit-pick: No need for a variable here, particularly as it's something that simple.
```suggestion
		switch_theme( 'block-theme-post-content-default' );

		$this->assertSame( array(), wp_get_post_content_block_attributes() );
```

**Review by felixarntz — CHANGES_REQUESTED** [review]
Thank you for the PR @tellthemachines. One thing looks off to me here as far as I can tell.

File: src/wp-includes/block-editor.php (hunk 1 above)
Comment (tellthemachines) [reply src/wp-includes/block-editor.php:639]: Oooh I see. The intention is for `$post_content_block_attributes` to be an empty array if the Post Content block exists but doesn't have any attributes set (that's the expected state of it when it only has default layout). 

I'm thinking the best solution would be to have `wp_get_post_content_block_attributes` return `null` by default instead of an empty array. The function is public though it's only been in core since 6.3; might that present any back compat problems? I did a quick search in wpdirectory and it didn't turn up anything.

Comment (tellthemachines) [reply src/wp-includes/block-editor.php:639]: Actually the function is marked private 😄 so it shouldn't be a problem. I made the change to the default return value.

File: tests/phpunit/tests/blocks/editor.php (hunk 2 above)
Comment (tellthemachines) [reply tests/phpunit/tests/blocks/editor.php:461]: The [original test](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/blocks/editor.php#L443) already checks the return value for classic themes (when Post Content doesn't exist), do you think it would be worth splitting that check out into its own test?

File: tests/phpunit/tests/blocks/editor.php [hunk 4]
```diff
@@ -447,13 +447,19 @@ public function test_wp_get_post_content_block_attributes() {
 			),
 		);
 		// With no block theme, expect an empty array.
-		$this->assertSame( array(), wp_get_post_content_block_attributes() );
+		$this->assertSame( null, wp_get_post_content_block_attributes() );
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/blocks/editor.php]:
```suggestion
		$this->assertNull( wp_get_post_content_block_attributes() );
```

Let's use `assertNull` instead `assertSame`

File: tests/phpunit/tests/blocks/editor.php [hunk 5]
```diff
@@ -447,13 +447,19 @@ public function test_wp_get_post_content_block_attributes() {
 			),
 		);
 		// With no block theme, expect an empty array.
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/blocks/editor.php]:
```suggestion
		// With no block theme, expect null.
```
Update docblock based on changes.

File: src/wp-includes/block-editor.php [hunk 6]
```diff
@@ -409,19 +409,20 @@ function wp_get_first_block( $blocks, $block_name ) {
  * Retrieves Post Content block attributes from the current post template.
  *
  * @since 6.3.0
+ * @since 6.4.0 return null by default and return empty array when Post Content block exists. but has no attributes.
```
Comment (mukeshpanchal27) [inline src/wp-includes/block-editor.php]:
```suggestion
 * @since 6.4.0 return null by default or an empty array when the Post Content block exists. but has no attributes.
```

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @tellthemachines for the PR. Left some nit-pick document and unit test changes.

File: tests/phpunit/tests/blocks/editor.php (hunk 5 above)
Comment (tellthemachines) [reply tests/phpunit/tests/blocks/editor.php]: Oooh thanks, totally missed that 😄

File: tests/phpunit/tests/blocks/editor.php (hunk 2 above)
Comment (felixarntz) [reply tests/phpunit/tests/blocks/editor.php:461]: @tellthemachines I meant a test for the `get_block_editor_settings()` change. As far as I can tell, there isn't one at the moment.

There is https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/blocks/editor.php#L524 which checks the presence of `postContentAttributes`, so I think we should add another test where we force `wp_get_post_content_block_attributes()` to return `null`, and then assert that `get_block_editor_settings()` does not include the `postContentAttributes` key.

File: src/wp-includes/block-editor.php [hunk 7]
```diff
@@ -409,19 +409,20 @@ function wp_get_first_block( $blocks, $block_name ) {
  * Retrieves Post Content block attributes from the current post template.
  *
  * @since 6.3.0
+ * @since 6.4.0 return null by default or an empty array when the Post Content block exists but has no attributes.
```
Comment (felixarntz) [inline src/wp-includes/block-editor.php]: I think we can simplify this to only explain the `null` change. The other part ("empty array when the Post Content block exists but has no attributes") was already like that before I'd argue.
```suggestion
 * @since 6.4.0 Return null if there is no post content block.
```

File: tests/phpunit/tests/blocks/editor.php (hunk 2 above)
Comment (tellthemachines) [reply tests/phpunit/tests/blocks/editor.php:461]: Oh got it! Yeah that's a good idea.

Comment (tellthemachines) [reply tests/phpunit/tests/blocks/editor.php:461]: Added a test in 0e75e53

**Review by felixarntz — APPROVED** [review]
Thank you for the updates @tellthemachines, LGTM!

**Comment by tellthemachines** [comment]
Thanks for the reviews! Committed in [r56629](https://core.trac.wordpress.org/changeset/56629).
